File: ckgame-java/src/main/java/com/ckgame/world/MapGraph.java
package com.ckgame.world;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** 地图：省份集合与邻接关系，支持 BFS 路径查找。 */
public final class MapGraph {
    private final Map<Integer, County> counties = new HashMap<>();
    private final Map<String, County> countyByKey = new HashMap<>();

    public void insert(County county) {
        counties.put(county.id, county);
        countyByKey.put(county.name.toLowerCase(), county);
    }

    public County get(int countyId) {
        return counties.get(countyId);
    }

    public County getByKey(String key) {
        return countyByKey.get(key.toLowerCase());
    }

    public void connect(int a, int b) {
        County ca = counties.get(a);
        County cb = counties.get(b);
        if (ca != null && !ca.neighbors.contains(b)) {
            ca.neighbors.add(b);
        }
        if (cb != null && !cb.neighbors.contains(a)) {
            cb.neighbors.add(a);
        }
    }

    public List<Integer> path(int from, int to) {
        if (from == to) {
            return List.of(from);
        }
        Set<Integer> visited = new HashSet<>();
        visited.add(from);
        Deque<Integer> queue = new ArrayDeque<>();
        queue.add(from);
        Map<Integer, Integer> parent = new HashMap<>();
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            County county = counties.get(cur);
            if (county == null) continue;
            for (int n : county.neighbors) {
                if (visited.contains(n)) continue;
                visited.add(n);
                parent.put(n, cur);
                if (n == to) {
                    List<Integer> path = new ArrayList<>();
                    path.add(to);
                    Integer p = n;
                    while (parent.containsKey(p)) {
                        int prev = parent.get(p);
                        path.add(prev);
                        if (prev == from) break;
                        p = prev;
                    }
                    java.util.Collections.reverse(path);
                    return path;
                }
                queue.add(n);
            }
        }
        return null;
    }

    public Iterator<County> iterator() {
        return counties.values().iterator();
    }

    /** 全部省份（无序）。 */
    public List<County> list() {
        return new ArrayList<>(counties.values());
    }

    public Map<Integer, County> counties() {
        return counties;
    }
}

File: ckgame-ts/src/world/MapGraph.ts
import { County } from './County.js';

/** 地图：省份集合与邻接关系，支持 BFS 路径查找。 */
export class MapGraph {
  private _counties: Map<number, County> = new Map();
  private countyByKey: Map<string, County> = new Map();

  insert(county: County): void {
    this._counties.set(county.id, county);
    this.countyByKey.set(county.name.toLowerCase(), county);
  }

  get(countyId: number): County | undefined {
    return this._counties.get(countyId);
  }

  getByKey(key: string): County | undefined {
    return this.countyByKey.get(key.toLowerCase());
  }

  connect(a: number, b: number): void {
    const ca = this._counties.get(a);
    const cb = this._counties.get(b);
    if (ca !== undefined && !ca.neighbors.includes(b)) {
      ca.neighbors.push(b);
    }
    if (cb !== undefined && !cb.neighbors.includes(a)) {
      cb.neighbors.push(a);
    }
  }

  path(from: number, to: number): number[] | null {
    if (from === to) {
      return [from];
    }
    const visited = new Set<number>();
    visited.add(from);
    const queue: number[] = [from];
    const parent = new Map<number, number>();

    while (queue.length > 0) {
      const cur = queue.shift()!;
      const county = this._counties.get(cur);
      if (county === undefined) continue;
      for (const n of county.neighbors) {
        if (visited.has(n)) continue;
        visited.add(n);
        parent.set(n, cur);
        if (n === to) {
          const result: number[] = [to];
          let p: number = n;
          while (parent.has(p)) {
            const prev = parent.get(p)!;
            result.push(prev);
            if (prev === from) break;
            p = prev;
          }
          result.reverse();
          return result;
        }
        queue.push(n);
      }
    }
    return null;
  }

  [Symbol.iterator](): Iterator<County> {
    return this._counties.values();
  }

  /** 全部省份（无序）。 */
  list(): County[] {
    return Array.from(this._counties.values());
  }

  /** 省份 Map 访问器。 */
  get counties(): Map<number, County> {
    return this._counties;
  }
}
